File: dotnet/Rezel/Grammars/ZigGrammar.cs
using Rezel.Common;
using Rezel.Lr;
using Rezel.Highlight;

namespace Rezel.Grammars;

public static class ZigGrammar
{
    private static readonly NodePropSource ZigHighlighting = HighlightUtil.StyleTags(new Dictionary<string, object>
    {
        ["asm enum fn struct test union var"] = Tags.DefinitionKeyword,
        ["comptime const continue defer errdefer export extern inline noalias noinline nosuspend pub resume"] = Tags.ModifierTag,
        ["addrspace align allowzero anyframe anytype callconv error linksection packed threadlocal unreachable volatile"] = Tags.ModifierTag,
        ["opaque"] = Tags.ModifierTag,
        ["if else switch for while case return break continue try"] = Tags.ControlKeyword,
        ["BlockLabel BreakLabel"] = Tags.LabelName,
        ["Identifier"] = Tags.VariableName,
        ["BuiltinIdentifier"] = Tags.Keyword,
        ["Name"] = Tags.Definition(Tags.VariableName),
        ["FnProto/Identifier"] = Tags.Function(Tags.Definition(Tags.VariableName)),
        ["VarDeclProto/TypeExpr/Identifier FnProto/TypeExpr/Identifier ContainerField/TypeExpr/Identifier ParamType/TypeExpr/Identifier"] = Tags.TypeName,
        ["AdditionOp"] = Tags.ArithmeticOperator,
        ["MultiplyOp"] = Tags.ArithmeticOperator,
        ["and or"] = Tags.LogicOperator,
        ["BitwiseOp"] = Tags.BitwiseOperator,
        ["BitShiftOp"] = Tags.BitwiseOperator,
        ["CompareOp"] = Tags.CompareOperator,
        ["AssignOp"] = Tags.DefinitionOperator,
        ["UpdateOp"] = Tags.UpdateOperator,
        ["ContainerDocComment"] = Tags.LineComment,
        ["DocComment"] = Tags.LineComment,
        ["LineComment"] = Tags.LineComment,
        ["Integer"] = Tags.Number,
        ["StringLiteral"] = Tags.String,
        ["StringLiteralSingle"] = Tags.String,
        ["( )"] = Tags.Paren,
        ["[ ]"] = Tags.SquareBracket,
        ["{ }"] = Tags.Brace,
        [".*"] = Tags.DerefOperator,
        [", ;"] = Tags.Separator,
    });

    private static LRParser? _parser;
    public static LRParser Parser => _parser ??= CreateParser();

    private static LRParser CreateParser()
    {
        var spec = ZigParserData.MakeSpec(
            propSources: [ZigHighlighting]
        );
        return new LRParser(spec);
    }
}

File: swift/Sources/Rezel/Grammars/ZigGrammar.swift
import Foundation

private nonisolated(unsafe) let definition = hlTags["definition"] as! (Tag) -> Tag
private nonisolated(unsafe) let functionMod = hlTags["function"] as! (Tag) -> Tag

nonisolated(unsafe) let zigHighlighting = styleTags([
	"asm enum fn struct test union var": hlTags["definitionKeyword"] as Any,
	"comptime const continue defer errdefer export extern inline noalias noinline nosuspend pub resume": hlTags["modifier"] as Any,
	"addrspace align allowzero anyframe anytype callconv error linksection packed threadlocal unreachable volatile": hlTags["modifier"] as Any,
	"opaque": hlTags["modifier"] as Any,
	"if else switch for while case return break continue try": hlTags["controlKeyword"] as Any,
	"BlockLabel BreakLabel": hlTags["labelName"] as Any,
	"Identifier": hlTags["variableName"] as Any,
	"BuiltinIdentifier": hlKeyword,
	"Name": definition(hlTags["variableName"] as! Tag),
	"FnProto/Identifier": functionMod(definition(hlTags["variableName"] as! Tag)),
	"VarDeclProto/TypeExpr/Identifier FnProto/TypeExpr/Identifier ContainerField/TypeExpr/Identifier ParamType/TypeExpr/Identifier": hlTypeName,
	"AdditionOp": hlTags["arithmeticOperator"] as Any,
	"MultiplyOp": hlTags["arithmeticOperator"] as Any,
	"and or": hlTags["logicOperator"] as Any,
	"BitwiseOp": hlTags["bitwiseOperator"] as Any,
	"BitShiftOp": hlTags["bitwiseOperator"] as Any,
	"CompareOp": hlTags["compareOperator"] as Any,
	"AssignOp": hlTags["definitionOperator"] as Any,
	"UpdateOp": hlTags["updateOperator"] as Any,
	"ContainerDocComment": hlTags["lineComment"] as Any,
	"DocComment": hlTags["lineComment"] as Any,
	"LineComment": hlTags["lineComment"] as Any,
	"Integer": hlNumber,
	"StringLiteral": hlString,
	"StringLiteralSingle": hlString,
	"( )": hlTags["paren"] as Any,
	"[ ]": hlTags["squareBracket"] as Any,
	"{ }": hlTags["brace"] as Any,
	".*": hlTags["derefOperator"] as Any,
	", ;": hlTags["separator"] as Any,
])

public let zigParser: LRParser = {
	let spec = ZigParserData.makeSpec(
		externals: [:],
		propSources: [zigHighlighting]
	)
	return LRParser(spec: spec)
}()
